tests: add parallel flavors of BenchmarkSysbench

tbg opened the pull request:
This PR adds a parallel flavor of BenchmarkSysbench. The implementation is 100% shared between the two, but a parallel flavor might be more adequate to demonstrate a throughput improvement for some changes, and is generally closer to what we see on roachprod clusters.
The single-threaded flavor continues to be useful as a low-variance measurement of end-to-end latency and allocations per operation.

Despite much internal refactoring, `BenchmarkSysbench`'s result remain ~unchanged:

<details><summary>BenchmarkSysbench</summary>
<p>

```
$ benchdiff --old=lastmerge ./pkg/sql/tests -b -r 'BenchmarkSysbench/SQL/3node/oltp_read_write.*' -d 3000x -c 20
old:  2e7d5a6 Merge #140477
new:  7fbc3d2 tests: extract BenchmarkParallelSysbench
args: benchdiff "--old=lastmerge" "./pkg/sql/tests" "-b" "-r" "BenchmarkSysbench/SQL/3node/oltp_read_write.*" "-d" "3000x" "-c" "20"

running benchmarks:
name                                   old time/op    new time/op    delta
Sysbench/SQL/3node/oltp_read_write-24    11.9ms ± 0%    11.9ms ± 0%   ~     (p=1.000 n=1+1)

name                                   old alloc/op   new alloc/op   delta
Sysbench/SQL/3node/oltp_read_write-24    2.34MB ± 0%    2.35MB ± 0%   ~     (p=1.000 n=1+1)

name                                   old allocs/op  new allocs/op  delta
Sysbench/SQL/3node/oltp_read_write-24     10.3k ± 0%     10.3k ± 0%   ~     (p=1.000 n=1+1)
```

</p>
</details> 

and the new parallel flavor gets around 10x the throughput (~900qps vs 84qps) while allocating less per op (likely due to pushing higher throughput and thus being able to batch background work more efficiently). I'll note that this is at 24 vCPUs, so under ideal scaling we would have expected ~24x the throughput. I don't think it is a realistic assumption because one goroutine worth of foreground work is tempered with background work, not to mention various possible inefficiencies in the system. It will be interesting to further study when linear scaling breaks down - see the mutex profile below.

<details><summary>BenchmarkParallelSysbench</summary>
<p>

```
$ benchdiff --old=HEAD ./pkg/sql/tests -b -r 'BenchmarkParallelSysbench/SQL/3node/oltp_read_write.*' -d 10000x -c 5
old:  7fbc3d2 tests: extract BenchmarkParallelSysbench
new:  7fbc3d2 tests: extract BenchmarkParallelSysbench
args: benchdiff "--old=HEAD" "./pkg/sql/tests" "-b" "-r" "BenchmarkParallelSysbench/SQL/3node/oltp_read_write.*" "-d" "10000x" "-c" "5"

name                                           old time/op    new time/op    delta
ParallelSysbench/SQL/3node/oltp_read_write-24    1.10ms ± 1%    1.10ms ± 1%   ~     (p=1.000 n=5+5)

name                                           old errs/op    new errs/op    delta
ParallelSysbench/SQL/3node/oltp_read_write-24      0.01 ± 6%      0.01 ± 6%   ~     (p=1.000 n=5+5)

name                                           old alloc/op   new alloc/op   delta
ParallelSysbench/SQL/3node/oltp_read_write-24    2.04MB ± 1%    2.04MB ± 1%   ~     (p=1.000 n=5+5)

name                                           old allocs/op  new allocs/op  delta
ParallelSysbench/SQL/3node/oltp_read_write-24     8.72k ± 1%     8.72k ± 1%   ~     (p=1.000 n=5+5)
```

</p>
</details> 

Profiles ([cpu.pb.gz](https://github.com/user-attachments/files/18690181/cpu.pb.gz) [mem.pb.gz](https://github.com/user-attachments/files/18690182/mem.pb.gz) [mutex.pb.gz](https://github.com/user-attachments/files/18690563/mutex.pb.gz)
) from

```
benchdiff --old=HEAD ./pkg/sql/tests -b -r 'BenchmarkParallelSysbench/SQL/3node/oltp_read_write.*'' -d 100000x -c 1 --cpuprofile --memprofile
```

This is a longer run than usual, but I wanted to drown out setting up the test cluster in the CPU profile and this did the trick.

<details><summary>CPU</summary>
<p>

![image](https://github.com/user-attachments/assets/7459ca6c-694f-49d4-abb5-f4de5d8f4e51)

</p>
</details> 

<details><summary>alloc_objects</summary>
<p>


![image](https://github.com/user-attachments/assets/7272217d-f75c-456d-9439-c8db15272991)


</p>
</details> 

<details><summary>alloc_space</summary>
<p>

![image](https://github.com/user-attachments/assets/694ff322-0583-4943-8dec-e5fed5aaddf7)


</p>
</details> 

The mutex profile is potentially interesting because it shows a very prominent contended mutex in `ssmemstorage.(*Container).getStatsForStmtWithKey`. I filed https://github.com/cockroachdb/cockroach/issues/140590 to discuss this.

<details><summary>Mutex</summary>
<p>

![image](https://github.com/user-attachments/assets/30b5cbdb-786f-455c-95c6-6444a96884be)


</p>
</details> 


Release note: none
Epic: CRDB-42584

tbg committed:
tests: extract sysbenchClient

tbg committed:
tests: in principle allow multiple sysbench clients

tbg committed:
tests: return errors from sysbench plumbing

This is in preparation for running concurrent txns, where retry
errors need to be dealt with.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/139557)
<!-- Reviewable:end -->

tbg committed:
tests: refactor sysbench test

Attempt to remove the indentation somewhat and to standardize
on a single implementation independent of the `parallel` param.

tbg committed:
tests: speed up BenchmarkSysbench

The Go harness has this annoying habit of invoking the benchmark first with
b.N==1, and only then with the (either prescribed or computed from benchtime)
desired number of iterations. But if we pass the iterations in, there's no
point in that, and the first invocation can be skipped, saving the very
significant setup cost (a few seconds a pop).

tbg committed:
tests: disable license enforcer in BenchmarkSysbench

In the parallel flavors of the benchmark, I saw it consume >2% CPU in a
profile. Lots of notices are being emitted. This is just not what real
world usage looks like, so disable the enforcer.

tbg committed:
tests: extract BenchmarkParallelSysbench

On second thought, it seems better to maintain BenchmarkSysbench's flavors as
is, and to add a new top-level BenchmarkParallelSysbench flavor.

That way, everyone's muscle memory keeps working (though folks might
accidentally run BenchmarkParallelSysbench as well, since the `Sysbench` regex
matches both) and also the microbench dashboard will not get confused due to
a change in metric names.

RaduBerinde reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/139557)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @tbg)

___
*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/139557#-OIQuRGi-wtQ9Dq7dANC:-OIQuRGi-wtQ9Dq7dAND:bg7igkv) (no related file):*
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 
___
*[`pkg/sql/tests/sysbench_test.go` line 856 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/139557#-OIQtI831WlGHOBMza09:-OIQtI831WlGHOBMza0A:b-j7nbx7) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f7c1742b00e16aaafed42978a3e383a6b7f5ebf/pkg/sql/tests/sysbench_test.go#L856)):*
> ```Go
> 				for _, workload := range workloads {
> 					b.Run(workload.name, func(b *testing.B) {
> 						defer func() {
> ```

[nit] this whole func could be separate, it's way too indented for a large block
___
*[`pkg/sql/tests/sysbench_test.go` line 863 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/139557#-OIQtxeY53SJSlQplh1e:-OIQtxeY53SJSlQplh1f:b-t6m5cx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff466290a6bebf635e1080711d1241b6281b5b8b/pkg/sql/tests/sysbench_test.go#L863)):*
> ```Go
> 
> 						var benchtime string
> 						require.NoError(b, sniffArgs(os.Args[1:], "test.benchtime", &benchtime))
> ```

go 1.24 will fix this https://tip.golang.org/doc/go1.24#new-benchmark-function :)
___
*[`pkg/sql/tests/sysbench_test.go` line 115 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139557#-OIQsh24DW8t14I5HJ88:-OIQsh24DW8t14I5HJ89:b-u9vvws) ([raw file](https://github.com/cockroachdb/cockroach/blob/75f011b27fcd8c0a8b122998100632403910d95c/pkg/sql/tests/sysbench_test.go#L115)):*
> ```Go
> 	DeleteInsert(tableNum, rowID, kValue, cValue, padValue) error
> 
> 	Close() error // destroys client, releasing any open txn
> ```

Rename to `Rollback()` and correct the comment
___
*[`pkg/sql/tests/sysbench_test.go` line 223 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/139557#-OIQsm_A9XotnZWpvWru:-OIQsm_A9XotnZWpvWrv:bsq05z9) ([raw file](https://github.com/cockroachdb/cockroach/blob/75f011b27fcd8c0a8b122998100632403910d95c/pkg/sql/tests/sysbench_test.go#L223)):*
> ```Go
> 
> 	conn *pgx.Conn
> 	open bool
> ```

`txnOpen` (I initially thought it's tracking whether a connection is open)


<!-- Sent from Reviewable.io -->

tbg committed:
tests: rename Close to Rollback and fix comment

tbg committed:
tests: rame sysbenchSQLClient.{open->txnOpen}

tbg committed:
sql: extract runSysbenchOuter

This avoids a big chunk of deeply indented code.

tbg committed:
tests: add note about b.Loop

tbg commented:
TFTR! All comments addressed.

bors r+
